rholang/src/main/scala/coop/rchain/rholang/interpreter/Substitute.scala

```diff
@@ -283,12 +291,16 @@ object Substitute {
               newLocallyFree = locallyFree.until(env.shift)
             } yield Expr(exprInstance = ETupleBody(ETuple(pss, newLocallyFree, connectiveUsed)))
 
-          case ESetBody(ESet(ps, locallyFree, connectiveUsed)) =>
+          case ESetBody(ParSet(SortedHashSet(ps), connectiveUsed, locallyFree)) =>
             for {
               pss <- ps.toList
                       .traverse(p => s1(p))
-              newLocallyFree = locallyFree.until(env.shift)
-            } yield Expr(exprInstance = ESetBody(ESet(pss, newLocallyFree, connectiveUsed)))
+            } yield
+              Expr(
+                exprInstance = ESetBody(
+                  ParSet(SortedHashSet(pss.toSeq),
+                         connectiveUsed,
+                         locallyFree.map(_.until(env.shift)))))
```

Comment:
@guardbotmk3 I used `Coeval(locallyFree` here and now it's possible to `map` it and go past already substituted values.

Reply:
Nice work.